### PR TITLE
Restart the async producer on errors

### PR DESCRIPTION
**Pull request by dasch**
This is an attempt to fix the cause of #422.

When there's an unexpected Kafka error, we should not crash the host application – that goes against what the async producer is there to do, which is insulate the host from Kafka problems.

Instead, we sleep for 10 seconds before restarting the background worker. This should both ensure that we don't burn crazy amounts of CPU if we run into a loop, but also that the user can look at the logs and take action.

### CODE REVIEW

File: lib/kafka/async_producer.rb [hunk 1]
```diff
@@ -218,6 +218,12 @@ def run
             raise "Unknown operation #{operation.inspect}"
           end
         end
+      rescue Kafka::Error => e
+        @logger.error "Unexpected Kafka error #{e.class}: #{e.message}\n#{e.backtrace.join("\n")}"
+        @logger.info "Restarting in 10 seconds..."
+
+        sleep 10
+        retry
```
Comment (film42) [inline lib/kafka/async_producer.rb]: Should we have a retry limit here in case the error is related to a problem with the message and not the connection? Example: https://github.com/abrandoned/protobuf-nats/blob/master/lib/protobuf/nats/client.rb#L96-L105

Comment (film42) [reply lib/kafka/async_producer.rb]: Also, 10 seconds might be a bit much. You could make this 1 second and it would still use a really low amount of system resources.

File: lib/kafka/async_producer.rb [hunk 2]
```diff
@@ -218,6 +218,12 @@ def run
             raise "Unknown operation #{operation.inspect}"
           end
         end
+      rescue Kafka::Error => e
+        @logger.error "Unexpected Kafka error #{e.class}: #{e.message}\n#{e.backtrace.join("\n")}"
```
Comment (film42) [inline lib/kafka/async_producer.rb]: 👍 to logging this error. Doesn't have to be now, but we should publish these errors out through the instrumenter.

File: lib/kafka/async_producer.rb [hunk 3]
```diff
@@ -218,6 +218,12 @@ def run
             raise "Unknown operation #{operation.inspect}"
           end
         end
+      rescue Kafka::Error => e
```
Comment (film42) [inline lib/kafka/async_producer.rb]: Just realized that you might want to make this `rescue => e` because it's still possible for some non-kafka error to crash the host while `thread.abort_on_exception = true` is still used.

File: lib/kafka/async_producer.rb (hunk 2 above)
Comment (dasch) [reply lib/kafka/async_producer.rb]: Yeah, it would make sense to track this in a dashboard.

File: lib/kafka/async_producer.rb (hunk 1 above)
Comment (dasch) [reply lib/kafka/async_producer.rb]: I think it's fairly typical to have a high number of e.g. Unicorn worker processes running on a host, and this would run per-worker – so if there's a systemic error, you'd get a restart per worker based on the frequency. I think 10s should suffice – this would only cover totally unforeseen errors, as we should add specific handling of errors whenever possible.

File: lib/kafka/async_producer.rb (hunk 3 above)
Comment (dasch) [reply lib/kafka/async_producer.rb]: That would be a bug that would warrant a hard crash.

File: lib/kafka/async_producer.rb (hunk 1 above)
Comment (film42) [reply lib/kafka/async_producer.rb]: Interesting. We run on jruby with puma backed by a single thread pool, so for us one instance has one async producer. Would you be open to making this configurable? I'm happy to make a follow-up change :)

File: lib/kafka/async_producer.rb (hunk 3 above)
Comment (film42) [reply lib/kafka/async_producer.rb]: Maybe for your workload but for us it's unacceptable. Would you be open to making this behavior configurable?

Comment (dasch) [reply lib/kafka/async_producer.rb]: Have you ever seen a non-Kafka error cause a crash? Do all other libraries you use rescue and hide all runtime errors?

Comment (mensfeld) [reply lib/kafka/async_producer.rb]: I'm with @dasch  on this one. @film42  can't you wrap around your logic with rescue?

Comment (dasch) [reply lib/kafka/async_producer.rb]: If you have very specific requirements for an async producer, it's also possible to create your own. The one included in ruby-kafka simply wraps the sync producer, so it's actually not that difficult.

Comment (film42) [reply lib/kafka/async_producer.rb]: @mensfeld I don’t have access to this code path because it’s in a separate thread that will crash the VM if something isn’t rescued. 

@dasch This is the first time I’ve seen a library that uses abort on exception to find bugs. I don’t think you need to handle any possible error ever here, but if you’re saying “any unhabdled error will crash the VM,” then I think that warrants more protection. But you’re right, it’s not hard to make my own. I was hoping we could find a way to make the default one more safe and generic. Sorry, I’m not trying to be annoying or argumentative. Thanks for working through this :)

So in summary I think not aborting on exception but publishing out all errors, and using a back-off timeout would cover both of our cases and expose any library bugs to the user. And I’m not asking you to write this, I’m just wondering if you think it’s a good pattern or not. I’m more than happy to make a cut and open a PR. I know you prefer the current implementation, but is there some kind of scenario where you’d be happy to remove abort on exception?

Comment (dasch) [reply lib/kafka/async_producer.rb]: I think not aborting on unexpected exceptions is reasonable, as long as the error is being logged. The worker thread would die though, so all writes would be dropped. 

Let me push an update to the PR.

Comment (film42) [reply lib/kafka/async_producer.rb]: I have some ideas on preserving writes between worker threads but as for now this branch performs the same as master branch (where the vm would stop and writes would be lost) so 👍  .

**Comment by film42** [comment]
Thanks @dasch !
